[apps] De-duplicate GSuite Events

stoggi opened the pull request:
to: @airbnb/streamalert-maintainers
cc: @ryandeivert 
size: medium
resolves #800

## Background

The startTime passed to the Google Reports API includes the last event already processed in the previous lambda execution. So duplicate events are observed in subsequent API polls.

## Changes

    * Created new key "context" stored in the state parameter in ssm
    * Save at most 50 event ids from the most recent timestamp in the context. Limit of 4096 characters in ssm.
    * Correctly format dates with a fractional second component.
    * De-duplicate events based on last seen event ids.
    * Update unit tests to output unique timestamps per generated log.
    * Added unit tests

## Testing

Added new unit tests:
* test_gather_logs_duplicate_events_more_to_poll
* test_gather_logs_duplicate_events_next_poll
* test_set_context_new
* test_set_context_same
* test_set_context_not_a_dictionary
* test_set_context_not_serializable

ryandeivert commented on stream_alert/apps/_apps/gsuite.py:
```diff
@@ -151,11 +161,24 @@ def _gather_logs(self):
+                for activity in activities
+                if activity['id']['time'] == self._last_timestamp
+            ]
+            self._context['next_event_ids'] = next_event_ids[:self._MAX_EVENT_IDS]
```
or just something similar using a local property that then caches to the instance prop if you want to keep the warning message below

stoggi replied:
The reason I needed to cache `_next_event_ids` and `_last_event_ids` is because the events are returned in reverse chronological order. The first few events contain the duplicates that we need to save, and the last events are the ones we want to de-duplicate. But in the case of a multiple data poll, there will be more than one invocation of `_gather_logs`, and the events at the end of the second poll will be the duplicates from the last polled time period, not necessarily the last lambda invocation. So we can't overwrite the cached event ids, and temporarily need to store both.

So, `_last_event_ids` are the event ids from the last poll, and `_next_event_ids` are the event ids for the next poll.

stoggi replied:
Refactored to only store one copy in the context since _gather_logs is called multiple times per execution. The case where it gets deferred to the next lambda execution is a very small edge case, and worst case you get some duplicate events.

ryandeivert commented on stream_alert/apps/_apps/gsuite.py:
```diff
@@ -151,11 +161,24 @@ def _gather_logs(self):
         self._more_to_poll = bool(self._next_page_token)
 
+        if not self._more_to_poll:
+            self._context['last_event_ids'] = self._context.get('next_event_ids', [])
```
this could then just save the cached instance property instead of copying a value that already exist in the context

stoggi replied:
It's true, at this point we no longer need to store `next_event_ids`. But only because there is nothing left to poll, and we know the next lambda invocation will be for a new time period. We could clear `next_event_ids`, but the important thing is to copy them across to `last_event_ids`.

stoggi replied:
Refactored to only store one copy in the context since `_gather_logs` is called multiple times per execution. The case where it gets deferred to the next lambda execution is a very small edge case, and worst case you get some duplicate events.

ryandeivert commented on stream_alert/apps/config.py:
```diff
@@ -273,9 +275,20 @@ def _determine_last_time(self, date_format):
+                self._CONTEXT_KEY: self.context,
+            })
+        except TypeError as err:
+            raise AppStateError('Could not serialise state for name \'{}\'. Error: '
```
I like the cultural diversity here :P but could we use the american english version of `serialise` here?

stoggi replied:
Ah yep, good point 👍

stoggi commented:
@ryandeivert addressed your questions, and made the following changes:

* Refactored to only store one copy of last_event_ids because _gather_logs is called many times per lambda execution.
* Increased the _MAX_EVENT_IDS since only one copy of last_event_ids is stored.
* Fixed typos.

ryandeivert commented:
thanks @stoggi !! I appreciate you following up on this :) I'm going to try to get to this soon (by EOW)
